mpt_trie/src/partial_trie.rs

```diff
@@ -155,301 +66,153 @@ where
-                },
-            ) => n1 == n2 && v1 == v2,
-            (_, _) => false,
-        }
-    }
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+#[serde(from = "Node", into = "Node")]
+pub struct FrozenNode {
+    node: Node,
+    hash: H256,
 }
 
-/// A simple PartialTrie with no hash caching.
-/// Note that while you can *still* calculate the hashes for any given node, the
-/// hashes are not cached and are recalculated each time.
-#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
-pub struct StandardTrie(pub Node<StandardTrie>);
-
-impl_from_for_trie_type!(StandardTrie);
-
-impl PartialTrie for StandardTrie {
-    fn new(n: Node<Self>) -> Self {
-        Self(n)
-    }
-
-    fn insert<K, V>(&mut self, k: K, v: V) -> TrieOpResult<()>
-    where
-        K: Into<Nibbles>,
-        V: Into<ValOrHash>,
-    {
-        self.0.trie_insert(k, v)?;
-        Ok(())
-    }
-
-    fn extend<K, V, I>(&mut self, nodes: I) -> TrieOpResult<()>
-    where
-        K: Into<Nibbles>,
-        V: Into<ValOrHash>,
-        I: IntoIterator<Item = (K, V)>,
-    {
-        self.0.trie_extend(nodes)
-    }
-
-    fn get<K>(&self, k: K) -> Option<&[u8]>
-    where
-        K: Into<Nibbles>,
-    {
-        self.0.trie_get(k)
-    }
-
-    fn delete<K>(&mut self, k: K) -> TrieOpResult<Option<Vec<u8>>>
-    where
-        K: Into<Nibbles>,
-    {
-        self.0.trie_delete(k)
-    }
-
-    fn hash(&self) -> H256 {
-        hash_trie(self)
-    }
-
-    fn items(&self) -> impl Iterator<Item = (Nibbles, ValOrHash)> {
-        self.0.trie_items()
-    }
-
-    fn keys(&self) -> impl Iterator<Item = Nibbles> {
-        self.0.trie_keys()
-    }
-
-    fn values(&self) -> impl Iterator<Item = ValOrHash> {
-        self.0.trie_values()
-    }
-
-    fn contains<K>(&self, k: K) -> bool
-    where
-        K: Into<Nibbles>,
-    {
-        self.0.trie_has_item_by_key(k)
+impl From<Node> for FrozenNode {
+    fn from(value: Node) -> Self {
+        value.freeze()
     }
 }
 
-impl TrieNodeIntern for StandardTrie {
-    fn hash_intern(&self) -> EncodedNode {
-        rlp_encode_and_hash_node(self)
+impl From<FrozenNode> for Node {
+    fn from(value: FrozenNode) -> Self {
+        value.thaw()
     }
 }
 
-impl Deref for StandardTrie {
-    type Target = Node<StandardTrie>;
-
-    fn deref(&self) -> &Self::Target {
-        &self.0
+impl Default for FrozenNode {
+    fn default() -> Self {
+        Node::default().freeze()
     }
 }
 
-impl DerefMut for StandardTrie {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.0
+impl FrozenNode {
+    pub fn hash(&self) -> H256 {
+        self.hash
     }
 }
 
-impl<K, V> TryFromIterator<(K, V)> for StandardTrie
-where
-    K: Into<Nibbles>,
-    V: Into<ValOrHash>,
-{
-    fn try_from_iter<T: IntoIterator<Item = (K, V)>>(nodes: T) -> TrieOpResult<Self> {
-        from_iter_common(nodes)
+impl FrozenNode {
+    pub fn thaw(self) -> Node {
```

Comment:
I had to google the meaning of `thaw` 😅 . Maybe worth explaining in the doc that we expose the inner node for non-native speakers